File: fm-dddrive/src/main/java/io/dddrive/core/property/model/EntityWithPropertiesSPI.java
package io.dddrive.core.property.model;

import org.springframework.lang.Nullable;

import io.dddrive.core.ddd.model.Part;
import io.dddrive.core.ddd.model.RepositoryDirectory;

public interface EntityWithPropertiesSPI {

	RepositoryDirectory getDirectory();

	boolean isInLoad();

	boolean isInCalc();

	Property<?> getParentProperty();

	String getRelativePath();

	String getPath();

	void fireEntityAddedChange(Object id);

	void fireEntityRemovedChange();

	void fireValueAddedChange(Property<?> property, Object value);

	void fireValueRemovedChange(Property<?> property, Object value);

	void fireFieldSetChange(Property<?> property, Object value, Object oldValue);

	void fireFieldChange(String op, String path, String value, String oldValue, boolean isInCalc);

	Part<?> doAddPart(Property<?> property, Integer partId);

	void doBeforeSet(Property<?> property, @Nullable Object value, @Nullable Object oldValue);

	void doAfterSet(Property<?> property);

	void doAfterClear(Property<?> property);

	void doAfterAdd(Property<?> property, Part<?> part);

	void doAfterRemove(Property<?> property);

}

File: fm-dddrive/src/main/java/io/dddrive/core/property/model/EntityWithPropertiesSPI.kt
package io.dddrive.core.property.model

import io.dddrive.core.ddd.model.Part
import io.dddrive.core.ddd.model.RepositoryDirectory

interface EntityWithPropertiesSPI {

	val directory: RepositoryDirectory

	val isInLoad: Boolean

	fun isInCalc(): Boolean

	val parentProperty: Property<*>?

	val relativePath: String

	val path: String

	fun fireFieldChange(
		op: String,
		path: String,
		value: String?,
		oldValue: String?,
		isInCalc: Boolean,
	)

	fun fireEntityAddedChange(id: Any)

	fun fireEntityRemovedChange()

	fun fireValueAddedChange(
		property: Property<*>,
		value: Any,
	)

	fun fireValueRemovedChange(
		property: Property<*>,
		value: Any,
	)

	fun fireFieldSetChange(
		property: Property<*>,
		value: Any?,
		oldValue: Any?,
	)

	fun doAddPart(
		property: Property<*>,
		partId: Int?,
	): Part<*>

	fun doBeforeSet(
		property: Property<*>,
		value: Any?,
		oldValue: Any?,
	)

	fun doAfterSet(property: Property<*>)

	fun doAfterClear(property: Property<*>)

	fun doAfterAdd(
		property: Property<*>,
		part: Part<*>?,
	)

	fun doAfterRemove(property: Property<*>)

	fun <T> getPropertyByPath(relativePath: String): Property<T>

}
